core/os/safe_binary_mutex.h

```diff
@@ -0,0 +1,124 @@
+/**************************************************************************/
+/*  mutex.h                                                               */
+/**************************************************************************/
+/*                         This file is part of:                          */
+/*                             GODOT ENGINE                               */
+/*                        https://godotengine.org                         */
+/**************************************************************************/
+/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#ifndef SAFE_BINARY_MUTEX_H
+#define SAFE_BINARY_MUTEX_H
+
+#include "core/error/error_macros.h"
+#include "core/typedefs.h"
```

Comment:
These two includes are unnecessary

Reply:
Transitive includes are a bad software engineering practice for multiple reasons:

- https://softwareengineering.stackexchange.com/a/262020/78524

In fact, the widespread use of this practice throughout the Godot codebase has made it terribly difficult for me to improve compilation times significantly, when it would have been quite simple otherwise. 

I recommend that every header and source file should be as explicit as possible with its `#include` dependencies, so that it can be compiled/refactored in isolation and does not depend on invisible chain of transitive dependencies. 

Since `DEV_ASSERT` and `_ALWAYS_INLINE` are respectively defined in `error_macros.h` and `typedefs.h`, and are directly used in this header file, they should be included explicitly. 

I hope you see the advantages of this approach but I am happy to discuss further if you have any doubts or concerns.

Reply:
Well that's how we do things, we don't include what we use, the discussion of this is separate and that's out of scope for this

Please follow how things are done, and discuss changing them separately 🙂

Reply:
> Well that's how we do things, we don't "include what we use", the discussion of this is separate and that's out of scope for this
> 
> Please follow how things are done, and discuss changing them separately 🙂
> 
> I'm well aware of the principle

From your own project's coding style guidelines (emphasis mine):

![{5645905F-53B9-431B-BF7E-0150001E4C42}](https://github.com/godotengine/godot/assets/918661/2ff9fb0f-bed4-4c32-a713-fe25dcf19a89)

As far as I can see, I am not breaking any rules, and I'm abiding to the last bullet point by "including what I use".

I also do not see any rule mentioning that I should rely on transitive includes.

Reply:
They're not written rules but they're rules we use, if also says "that requires them" which isn't the same

But trust me this is a rule, even if it's not written, let's not get bogged down in the letter of the law (though that page could do with an update)

If you want a second opinion I can provide it, but let's just remove these superfluous includes here

Reply:
As I strongly feel this is a mistake, I'd like to start a discussion/issue and get an official statement on the matter after a technical discussion.

What is the appropriate forum to start such discussion? Should I open an issue here, or should I open a thread on `godot-proposals`?

Reply:
A proposal on godot-proposals, as an issue and not a proposal with the relevant details 🙂

Reply:
Note that this was explicitly rejected before, see:
* https://github.com/godotengine/godot/pull/57760#issuecomment-1503228157

Reply:
But if you'll remove these two includes we can run the CI for this and see, and leave the discussion for later

Reply:
@AThousandShips: created a discussion here as requested: https://github.com/godotengine/godot-proposals/discussions/9016

Reply:
I said an issue with larger details but a discussion is fine too

Reply:
> I said an issue with larger details but a discussion is fine too

I apologize, I started writing the discussion after you said

> If you feel very strongly about "include what you use" please open a [proposal](https://github.com/godotengine/godot-proposals) if one doesn't already exist and advocate for it

and missed the new comment where you clarified that it should have been an issue.

Reply:
I'm sorry for taking a bit harder stance on this than I intended to come across, but I felt this was a simple and non-controversial suggestion following what I'm used to in the codebase, and didn't expect it to become a huge deal to even argue over

Will let someone else review this and hope we can deal with this maturely and professionally and not take these kinds of differences personally 🙂

Have a good day